### DISCUSSION TITLE
Help displaying info of selected track on OLED

### DISCUSSION THREAD

**Discussion by ACuppaJoes** [tttapa/Control-Surface#545] [Q&A]
Hey Pieter, 

I am trying to build an informational display like the one used in the example MCU-OLED-SSD1306-x2.ino, but instead of displaying information of tracks within a bank, I would like a single i2c OLED screen to display information of just the selected track. After modifying code from the example I am facing this error. "expected primary-expression before 'timedisplaydisplay'"

When I compile the example, everything is fine. So it must be an issue with how I am attempting to remove the use of banks.

Any chance you could help me out? Code is below

Code:
```
// ----------------------------- Main Libraries ----------------------------- //
// ========================================================================== //

#include <Control_Surface.h>
//#include <menu.h>

// -------------------------- Peripheral Libraries -------------------------- //
// ========================================================================== //

//#include <menuIO/keyIn.h> //Button inputs
//#include <menuIO/chainStream.h>
//#include <menuIO/serialIO.h>

// --------------------------- Display Libraries ---------------------------- //
// ========================================================================== //

#include <SPI.h>
#include <Wire.h>
#include <Adafruit_SSD1306.h>
#include <Display/DisplayInterface.hpp>
#include <Display/DisplayInterfaces/DisplayInterfaceSSD1306.hpp>


// ----------------------------- MIDI Interface ----------------------------- //
// ========================================================================== //

USBMIDI_Interface midi;
//USBDebugMIDI_Interface midi(115200);
int MidiCH = 1; //Midi channel number 1-16

// --------------------------- Display interface ---------------------------- //
// ========================================================================== //


#define SCREEN_WIDTH 128 // OLED display width, in pixels
#define SCREEN_HEIGHT 64 // OLED display height, in pixels

#define OLED_RESET     4 // Reset pin # (or -1 if sharing Arduino reset pin)
#define SCREEN_ADDRESS 0x3D ///< See datasheet for Address; 0x3D for 128x64, 0x3C for 128x32
Adafruit_SSD1306 OLEDdisplay(SCREEN_WIDTH, SCREEN_HEIGHT, &Wire, OLED_RESET);

// -------------------------- MIDI Input Elements --------------------------- //
// ========================================================================== //

/*
   Define all elements that listen for MIDI messages.
*/

// Main MCU LCD screen, used to get track names
MCU::LCD<> lcd {};

// Time display keeps track of the bar counter
MCU::TimeDisplay timedisplay {};

// Play / Record
NoteValue play {MCU::PLAY};
NoteValue record {MCU::RECORD};

// Record arm / ready
NoteValue recrdy {MCU::REC_RDY_1};

// VU meters
MCU::VU vu {MCU::VUDecay::Hold};

// ---------------------------- Display Elements ---------------------------- //
// ========================================================================== //

/*
   Define all display elements that display the state of the input elements.
*/


// Track names
MCU::LCDDisplay lcddisps {
  // track (1), position (0, 40), font size (1)
  {OLEDdisplay, lcd, {0, 40}, 1, WHITE},

// Time display
MCU::TimeDisplayDisplay timedisplaydisplay {
  // position (0, 0), font size (1)
  OLEDdisplay, timedisplay, {0, 0}, 1, WHITE,
};

// Play / Record
BitmapDisplay<> playDisp {
  OLEDdisplay, play, XBM::play_7, {16 + 64, 0}, WHITE,
};
BitmapDisplay<> recordDisp {
  OLEDdisplay, record, XBM::record_7, {26 + 64, 0}, WHITE,
};

// Record arm / ready
BitmapDisplay<> recrdyDisp {
  {OLEDdisplay, recrdy, XBM::rec_rdy_10B, {14 + 14, 50}, WHITE},
};

// VU meters
MCU::VUDisplay<> vuDisp {
  // position (32+11, 60), width (16), bar height (3) px, bar spacing (1) px
  {OLEDdisplay, vu, {32 + 11, 60}, 16, 3, 1, WHITE},
};

// --------------------------------- Setup ---------------------------------- //
// ========================================================================== //

void setup() {
  Control_Surface.begin();
}
// ---------------------------------- Loop ---------------------------------- //
// ========================================================================== //

void loop() {
  Control_Surface.loop(); // Refresh all elements
 
}

```

**Comment by tttapa** [comment]
> `MCU::LCDDisplay lcddisps {`

This bracket is never closed.

**Reply by ACuppaJoes** [reply]
Even after closing it, I get this error "no matching function for call to 'CS::MCU::LCDDisplay::LCDDisplay(<brace-enclosed initializer list>)'"

**Reply by tttapa** [reply]
The original example initialized an array of LCDDisplay objects, you're initializing a single object, so you only need a single pair of braces. 

```cpp
MCU::LCDDisplay lcddisps {
  OLEDdisplay, lcd, {0, 40}, 1, WHITE,
};
```
You also seem to be missing the `track` constructor argument, see https://tttapa.github.io/Control-Surface-doc/Doxygen/d7/dae/classMCU_1_1LCDDisplay.html#a62090ad620f0bf9e581b475624ae9d34.

**Reply by ACuppaJoes** [reply]
Even after only using a single pair of braces, and adding the track constructor argument, I am still getting the same error 

[no matching function for call to 'CS::MCU::LCDDisplay::LCDDisplay(<brace-enclosed initializer list>)']

To my understanding, this is how the code should look.

MCU::LCDDisplay lcddisps {
  OLEDdisplay, lcd, 1, {0, 40}, 1, WHITE
};

**Reply by tttapa** [reply]
I should have looked at your code more thoroughly: the first argument is a `DisplayInterface`, not `Adafruit_SSD1306`. See the examples on how to define the display interface.

**Reply by ACuppaJoes** [reply]
I have also included the <Display/DisplayInterfaces/DisplayInterfaceSSD1306.hpp> like the example. The reason I also included <Adafruit_SSD1306.h> is for use with a second library called Arduino Menu. 

I just tried commenting out the other display libraries and still face the same error

#include <Wire.h>
//#include <Adafruit_SSD1306.h>
//#include <Display/DisplayInterface.hpp>
#include <Display/DisplayInterfaces/DisplayInterfaceSSD1306.hpp>

I have also tried only including <Wire.h> and <Display/DisplayInterface.hpp>......same error

**Reply by tttapa** [reply]
Including it is insufficient, you have to define your own display interface and initialize it. Then you pass this interface to the display elements, not the Adafruit display directly.

https://github.com/tttapa/Control-Surface/blob/485d89ab347110e018cf86f03d470046e4f90d09/examples/7.Displays/MCU-OLED-SSD1306/MCU-OLED-SSD1306.ino#L87-L108

**Reply by ACuppaJoes** [reply]
Yup, that did the trick! Thanks. So does this mean that I can not use the [Arduino Menu library](https://github.com/neu-rah/ArduinoMenu) alongside <Display/DisplayInterfaces/DisplayInterfaceSSD1306.hpp>?

In some of the examples, they have used the Adafruit_SSD1306 library. I noticed DisplayInterfaces is also based on Adafruit_SSD1306. I would really like the ability to add a menu that allows me to select presets and change midi channel. These presets would let me change what midi notes are sent from pressing buttons 1,2 and 3 as well as the channel they are sent on.

**Reply by ACuppaJoes** [reply]
Also do you have any suggestions on how I can reduce program size? Unfortunately, this program is too big to run on a Leonardo board. I have already reduced the sysex buffer size to 64 in the settings. Is there anything else I can do?

I just ran a test and only included <Control_Surface.h>, and the library takes up 50% of the program storage. Can I use something like <Display/DisplayInterfaces/DisplayInterfaceSSD1306.hpp> to direct the program only to specific control surface components that are needed?

**Reply by tttapa** [reply]
> So does this mean that I can not use the [Arduino Menu library](https://github.com/neu-rah/ArduinoMenu) alongside <Display/DisplayInterfaces/DisplayInterfaceSSD1306.hpp>?

Not necessarily, but you have to synchronize them in some way: you cannot have two pieces of code writing to the same display buffer and using the same physical display without some coordination.

The way Control Surface handles displays is as follows: when you call `Control_Surface.loop()`,
1. It asks all display elements if they have any data to draw: Control Surface calls the `DisplayElement::getDirty()` function for each element, these return true if the state is “dirty” (i.e. the actual data on the display doesn't match the data that should be displayed).
2. If none of the elements were “dirty”, nothing needs to be done.
3. If at least one element wants to draw something to a display, the following happens:
  a. The display buffer is cleared by calling `DisplayInterface::clearAndDrawBackground()`
  b. All display elements on that display are iterated over, and `DisplayElement::draw()` is called for each of them, giving them the opportunity to draw anything they want to the display buffer.
  c. The display buffer is sent to the display by calling the `DisplayInterface::display()` function. In this case, this function will just call the `Adafruit_SSD1306::display()` function that you may be familiar with.

The menu library will probably call `Adafruit_SSD1306::display()` under the hood, so you'll have to find some way to integrate that with Control Surface, I'm not familiar with that menu library, so I'm afraid I can't really help you with that.

An alternative is not to use the Control Surface display system at all: you just use the MIDI Input Elements like `MCU::LCD`, without using the Display Elements like `MCU::LCDDisplay`. This means that you have to draw all information to the display manually, somewhere in your menu system.

> Also do you have any suggestions on how I can reduce program size? Unfortunately, this program is too big to run on a Leonardo board.

That's not easy. The Leonardo doesn't have a lot of memory, and things like displays require tons of memory.  
Control Surface itself requires quite a bit of memory as well, there's no way around that: there's a trade-off between ease of use and memory usage, and I decided that the former was more important. Most modern Arduino boards have much more memory and processing power, and are often cheaper or the same price as the Arduino Uno, Leonardo, etc., so sacrificing usability just to support underpowered 8-bit AVR boards would not be a good strategy IMHO.

I usually recommend Teensy boards for MIDI controllers with displays etc., they cost about the same as an Arduino Leonardo, and have orders of magnitude more memory (2MiB flash and 1MiB RAM, vs 32KiB flash and 2.5KiB RAM on the Leonardo).
  
If you were to write the Control Surface functionality from scratch, with focus on memory usage rather than user friendliness as a library, you could probably fit it on an Arduino Leonardo, but realistically, you're much better off just using a microcontroller that has enough memory for the job. Spending weeks rewriting all code is not worth it compared to just buying a suitable Arduino.

> I have already reduced the sysex buffer size to 64 in the settings.

This might cause the `MCU::LCD` to stop working. The SysEx messages sent to update the display are usually well over 100 bytes in length. If the buffer is only 64 bytes, Control Surface has to cut them up in chunks, and this would require extra logic and persistent state in the `MCU::LCD` implementation, so `MCU::LCD` only supports complete messages, not chunks.

> I just ran a test and only included <Control_Surface.h>, and the library takes up 50% of the program storage. Can I use something like <Display/DisplayInterfaces/DisplayInterfaceSSD1306.hpp> to direct the program only to specific control surface components that are needed?

I'll see what I can do, but I'm afraid it's not much. If you're using MIDI input and displays, you're using most parts of the library anyway.

**Reply by ACuppaJoes** [reply]
Thanks for the info. It seems Leonardo is a bad choice for this project. Once I finish the prototype, I may choose to turn it into a product. Unfortunately, the teensy would add 50% to the overall component cost. Do you have any other board suggestions? perhaps the arduino mega pro mini with a USBMidiKliK firmware flash? or will that flash be incompatible with control surface and a display?

if not, I may have to just bite the bullet and go for teensy 3.2

**Reply by tttapa** [reply]
Teensies are great for hobby projects, but for a commercial product, the price is of course something to keep in mind. PJRC offers options to buy the chips and the bootloader separately, which might be cheaper: https://www.pjrc.com/store/ic_mkl02.html  

An alternative could be the SAMD21 chip used in the Nano 33 IoT, for example. I haven't tested it personally, but all Control Surface examples seem to compile without issues, using the official Arduino USBMIDI library. There are also other non-Arduino SAMD21-based development boards.

I'm not a huge fan of the 8-bit AVR boards, so I wouldn't recommend a Mega either, I think ARM boards offer much better value. USBMidiKliK works okay but is a huge bottleneck by using serial communication, I'd highly recommend getting a chip with native USB MIDI support.

**Reply by ACuppaJoes** [reply]
I never knew they sold pre-flashed chips. The price is pretty reasonable as well. I think that might be the best option moving forward

**Reply by tttapa** [reply]
> The price is pretty reasonable as well.

Keep in mind that that's just the price for the bootloader chip, you need to buy the main microcontroller separately. (And it looks like the MK20DX256VLH7 might be out of stock at most suppliers, given the global chip shortage.)

**Reply by ACuppaJoes** [reply]
Hey Pieter,

Not sure what's going on but when I try to dictate the communication type using USBMIDI_Interface midicom; my code compiles perfectly fine in the Arduino IDE but when I copy the same code to the Teensyduino IDE, I get the error "'USBMIDI_Interface' does not name a type". It's the same code as above, I even tried changing the name from midi to midicom. I'm not sure if this is an issue with the library or something I have done.

Update: I just tried commenting out USBMIDI_Interface midicom; and used USBDebugMIDI_Interface midi(115200); instead and it compiles fine

**Reply by tttapa** [reply]
Did you select “USB Type > MIDI” in the Tools menu of the IDE?

[![Teensyduino USB Type menu](https://www.pjrc.com/teensy/teensyduino_sc1_mini.png)](https://www.pjrc.com/teensy/teensyduino.html)
